quinn-proto/src/connection/paths.rs

```diff
@@ -0,0 +1,186 @@
+            congestion,
+            mtud: MtuDiscovery::new(remote),
+        }
+    }
+
+    pub fn from_previous(remote: SocketAddr, prev: &PathData) -> Self {
+        PathData {
+            remote,
+            rtt: prev.rtt,
+            congestion: prev.congestion.clone_box(),
+            sending_ecn: true,
+            mtud: MtuDiscovery::new(remote),
+        }
+    }
+}
+
+#[derive(Copy, Clone)]
+pub struct RttEstimator {
+    /// The most recent RTT measurement made when receiving an ack for a previously unacked packet
+    latest: Duration,
+    /// The smoothed RTT of the connection, computed as described in RFC6298
+    smoothed: Option<Duration>,
+    /// The RTT variance, computed as described in RFC6298
+    var: Duration,
+    /// The minimum RTT seen in the connection, ignoring ack delay.
+    min: Duration,
+}
+
+impl RttEstimator {
+    fn new() -> Self {
+        Self {
+            latest: Duration::new(0, 0),
+            smoothed: None,
+            var: Duration::new(0, 0),
+            min: Duration::new(u64::max_value(), 0),
+        }
+    }
+
+    pub fn update(&mut self, ack_delay: Duration, rtt: Duration) {
+        self.latest = rtt;
+        // min_rtt ignores ack delay.
+        self.min = cmp::min(self.min, self.latest);
+        // Adjust for ack delay if it's plausible.
+        if self.latest - self.min > ack_delay {
+            self.latest -= ack_delay;
+        }
+        // Based on RFC6298.
+        if let Some(smoothed) = self.smoothed {
+            let var_sample = if smoothed > self.latest {
+                smoothed - self.latest
+            } else {
+                self.latest - smoothed
+            };
+            self.var = (3 * self.var + var_sample) / 4;
+            self.smoothed = Some((7 * smoothed + self.latest) / 8);
+        } else {
+            self.smoothed = Some(self.latest);
+            self.var = self.latest / 2;
+        }
+    }
+
+    pub fn get(&self) -> Duration {
+        self.smoothed
+            .map_or(self.latest, |x| cmp::max(x, self.latest))
+    }
+
+    pub fn pto_base(&self) -> Option<Duration> {
+        self.smoothed
+            .map(|srtt| srtt + cmp::max(4 * self.var, TIMER_GRANULARITY))
+    }
+}
+
+// Implements Datagram Packetization Layer Path Maximum Transmission Unit Discovery
+//
+// https://www.ietf.org/id/draft-ietf-tsvwg-datagram-plpmtud-21.html
+pub struct MtuDiscovery {
+    header_size: u16,
+    // Current MTU for the path
+    pub current: u16,
+    // Packet number and probe size for the current probe
+    probe_number: Option<u64>,
+    probe_size: Option<u16>,
+    // Failed probes at the current probe size
+    probe_count: usize,
+    phase: Phase,
+}
+
+impl MtuDiscovery {
+    fn new(remote: SocketAddr) -> Self {
+        Self {
+            header_size: match remote {
+                SocketAddr::V4(_) => 20,
+                SocketAddr::V6(_) => 48,
+            },
+            current: BASE_PLPMTU,
```

Comment:
`BASE_PLPMTU` is 1280, which is the [documented](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-packet-size) minimum *IP* packet size, so it's incorrect to add header sizes it or use it as the maximum UDP payload size.

Could we ditch the `header_size` field entirely, and just initialize the current state to `1280 - match ...`?